radix-engine-tests/tests/db/jmt_consistency.rs

```diff
@@ -1,37 +1,30 @@
 use radix_common::prelude::*;
-use radix_transaction_scenarios::scenario::{NextAction, ScenarioCore};
-use radix_transaction_scenarios::scenarios::get_builder_for_every_scenario;
+use radix_substate_store_impls::state_tree_support::*;
+use radix_transaction_scenarios::scenarios::*;
 use scrypto_test::prelude::*;
 
 #[test]
 fn substate_store_matches_state_tree_after_each_scenario() {
-    let network = NetworkDefinition::simulator();
-    let mut ledger = LedgerSimulatorBuilder::new().with_state_hashing().build();
-
-    let mut next_nonce: u32 = 0;
-    for scenario_builder in get_builder_for_every_scenario() {
-        let epoch = ledger.get_current_epoch();
-        let mut scenario = scenario_builder(ScenarioCore::new(network.clone(), epoch, next_nonce));
-        let mut previous = None;
-        loop {
-            let next = scenario
-                .next(previous.as_ref())
-                .map_err(|err| err.into_full(&scenario))
-                .unwrap();
-            match next {
-                NextAction::Transaction(next) => {
-                    let receipt = ledger.execute_notarized_transaction(&next.raw_transaction);
-                    // TODO(when figured out): We could run this assert as part of the existing
-                    // `post_run_db_check` feature; however, it is tricky to implement (syntactically),
-                    // due to the assert only being available for certain `TestDatabase` impl.
-                    ledger.assert_state_tree_matches_substate_store();
-                    previous = Some(receipt);
-                }
-                NextAction::Completed(end_state) => {
-                    next_nonce = end_state.next_unused_nonce;
-                    break;
-                }
+    let db = StateTreeUpdatingDatabase::new(InMemorySubstateDatabase::standard());
+    DefaultTransactionScenarioExecutor::new(db)
+        .on_transaction_execution(|_, _, _, db| {
+            let hashes_from_tree = db.list_substate_hashes();
```

Comment:
Hmm we have duplicated the content of `assert_state_tree_matches_substate_store` here, which is a little unclean.

Probably we still want `LedgerSimulator`'s `assert_state_tree_matches_substate_store` but maybe they should call into some common method or something?

Reply:
Yeah agreed that it would be better to have just a single method for this check. I've refactored this into a `validate_state_tree_matches_substate_store` method on the `StateTreeUpdatingDatabase<D>` type that returns a `Result`. In tests this result is not used in any meaningful way and is just unwrapped. The main reason for having it be a `Result` and not a `panic` or an `assert` is to avoid having any panicing APIs on DB types.